### DISCUSSION TITLE
Expose function args from high order function

### DISCUSSION THREAD

**Discussion by Tomperez98** [microsoft/pyright#7910] [Q&A] [answered]
```py
from __future__ import annotations

from typing import Callable

from typing_extensions import ParamSpec, TypeVar

P = ParamSpec("P")
T = TypeVar("T")


def run(fn: Callable[P, T]) -> T:
    return fn()


def greet(name: str) -> str: ...


def main():
    run(greet, name="tomas") # Is there a way to get type annotations for the name argument?
```

Is it possible to have typing annotation for the `greet` arguments when function being called from `run`? I need to define a high order function that received a function points and its arguments separately.

**Comment by erictraut** [comment, marked as answer]
I don't know what you mean by "get the type annotations for the name argument". 

If you want `run` to accept arguments that it can pass to `fn`, you need to include `*args` and `**kwargs` parameters like this:

```python
def run(fn: Callable[P, T], *args: P.args, **kwargs: P.kwargs) -> T:
    return fn(*args, **kwargs)
```

**Reply by Tomperez98** [reply]
I wanted example that. To get an error when calling `run(greet)` requiring me to add `name` as a parameter.

**Reply by Tomperez98** [reply]
All the functions the `run` functions will run required to have a `Context` argument

```py
from __future__ import annotations

from typing import Callable

from typing_extensions import ParamSpec, TypeVar, Concatenate

class Context: ...
T = TypeVar("T")
P = ParamSpec("P")


def run(fn: Callable[Concatenate[Context, P],  T], *args: P.args, **kwargs: P.kwargs) -> T:
    return fn(*args, **kwargs)


def greet(ctx: Context, name: str) -> str: ...


x = run(greet, name="tomas")

```

What is the correct way to use `Concatenate` to define this requirement and the pass all the extra args to the function?
